connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala

```diff
@@ -0,0 +1,246 @@
+import scala.util.control.NonFatal
+
+import org.apache.spark.connect.proto
+import org.apache.spark.connect.proto.{AddArtifactsRequest, AddArtifactsResponse}
+import org.apache.spark.connect.proto.AddArtifactsResponse.ArtifactSummary
+import org.apache.spark.sql.connect.artifact.SparkConnectArtifactManager
+import org.apache.spark.util.Utils
+
+/**
+ * Handles [[AddArtifactsRequest]]s for the [[SparkConnectService]].
+ *
+ * @param responseObserver
+ */
+class SparkConnectAddArtifactsHandler(val responseObserver: StreamObserver[AddArtifactsResponse])
+    extends StreamObserver[AddArtifactsRequest] {
+
+  // Temporary directory where artifacts are rebuilt from the bytes sent over the wire.
+  protected val stagingDir: Path = Utils.createTempDir().toPath
+  protected val stagedArtifacts: mutable.Buffer[StagedArtifact] =
+    mutable.Buffer.empty[StagedArtifact]
+  // If not null, indicates the currently active chunked artifact that is being rebuilt from
+  // several [[AddArtifactsRequest]]s.
+  private var chunkedArtifact: StagedChunkedArtifact = _
+  private var holder: SessionHolder = _
+  private def artifactManager: SparkConnectArtifactManager =
+    SparkConnectArtifactManager.getOrCreateArtifactManager
+
+  override def onNext(req: AddArtifactsRequest): Unit = {
+    if (this.holder == null) {
+      val holder = SparkConnectService.getOrCreateIsolatedSession(
+        req.getUserContext.getUserId,
+        req.getSessionId)
+      this.holder = holder
+    }
+
+    if (req.hasBeginChunk) {
+      // The beginning of a multi-chunk artifact.
+      require(chunkedArtifact == null)
+      chunkedArtifact = writeDepToFile(req.getBeginChunk)
+    } else if (req.hasChunk) {
+      // We are currently processing a multi-chunk artifact
+      require(chunkedArtifact != null && chunkedArtifact.getRemainingChunks > 0)
+      chunkedArtifact.write(req.getChunk)
+
+      if (chunkedArtifact.isFinished) {
+        chunkedArtifact.close()
+        // Unset the currently active chunked artifact.
+        chunkedArtifact = null
+      }
+    } else if (req.hasBatch) {
+      req.getBatch.getArtifactsList.forEach { artifact =>
+        // Each artifact in the batch is single-chunked.
+        val out = writeDepToFile(artifact)
+        out.close()
+      }
+    } else {
+      throw new UnsupportedOperationException(
+        s"$req could not be processed due to unsupported" +
+          s" data transfer mechanism")
+    }
+  }
+
+  override def onError(throwable: Throwable): Unit = {
+    Utils.deleteRecursively(stagingDir.toFile)
+    responseObserver.onError(throwable)
+  }
+
+  protected def addStagedArtifactToArtifactManager(artifact: StagedArtifact): Unit = {
+    artifactManager.addArtifact(holder.session, artifact.path, artifact.stagedPath)
+  }
+
+  /**
+   * Process all the staged artifacts built in this stream.
+   *
+   * @return
+   */
+  protected def flushStagedArtifacts(): Seq[ArtifactSummary] = {
+    stagedArtifacts.map { artifact =>
```

Comment:
You need to be a bit careful with this kind of code. When operating on a Seq `map` can actually be lazy. In this case you are fine because you are using a buffer.

Reply:
Ack. Left a comment in the method.